### PR TITLE
remove vscode ui wrap, add icon before create new server label

### PR DESCRIPTION
**Pull request by SummerSun**
vsode builds a very good show dialog ui, remove the VScodeUI wrap to reduce cost.

### CODE REVIEW

**Comment by SummerSun** [comment]
#54

File: src/Tomcat/TomcatServer.ts [hunk 1]
```diff
@@ -15,10 +19,11 @@ export class TomcatServer {
     private _isDebugging: boolean = false;
 
     constructor(name: string, tomcatPath: string, extensionPath: string) {
-        this._name = name.trim();
-        this._tomcatPath = tomcatPath.trim();
+        this._name = name;
+        this.label = name;
+        this._tomcatPath = tomcatPath;
```
Comment (andxu) [inline src/Tomcat/TomcatServer.ts]: use TypeScript Constructor Assignment to reduce a lot of codes. https://kendaleiv.com/typescript-constructor-assignment-public-and-private-keywords/

File: src/Tomcat/TomcatServer.ts [hunk 2]
```diff
@@ -4,8 +4,12 @@ import * as path from "path";
 import * as vscode from "vscode";
 import { Utility } from "../Utility";
 
-export class TomcatServer {
+export class TomcatServer implements vscode.QuickPickItem {
```
Comment (andxu) [inline src/Tomcat/TomcatServer.ts]: rename TomcatServer to TomcatServerItem

File: src/extension.ts [hunk 3]
```diff
@@ -136,16 +113,23 @@ async function openServerConfig(tomcat: TomcatController, tomcatItem ?: TomcatSe
...
 async function createServer(tomcat: TomcatController): Promise<string> {
-    const ui: VSCodeUI = new VSCodeUI();
-    const tomcatPath: string = await ui.showFileFolderDialog(false, true, DialogMessage.selectDirectory);
-    const serverName: string = path.basename(tomcatPath);
-
-    if (tomcat.getTomcatServer(serverName)) {
-        vscode.window.showInformationMessage(DialogMessage.serverExist);
-    } else {
-        await tomcat.createTomcatServer(serverName, tomcatPath);
+    const pathPick: vscode.Uri[] = await vscode.window.showOpenDialog(
```
Comment (andxu) [inline src/extension.ts]: bad identation

File: src/Tomcat/TomcatServer.ts (hunk 2 above)
Comment (SummerSun) [reply src/Tomcat/TomcatServer.ts]: Mostly use of TomcatServer is its server feature. Or, we could make a independent class TomcatServerQuickPickItem, but it seems not that necessary, just use TomcatServer.

File: src/Tomcat/TomcatServer.ts [hunk 4]
```diff
@@ -5,21 +5,16 @@ import * as vscode from "vscode";
...
-        this._name = name.trim();
-        this._tomcatPath = tomcatPath.trim();
-        this._extensionPath = extensionPath.trim();
-    }
+export class TomcatServer implements vscode.QuickPickItem {
+    // tslint:disable:no-parameter-properties no-unnecessary-initializer *
+    constructor(private _name: string, private _tomcatPath: string, private _extensionPath: string,
+                private _state: ServerState = ServerState.IdleServer, private _started: boolean = false,
+                private _isDebugging: boolean = false, private _debugPort: number = undefined,
+                private _debugWorkspace: vscode.WorkspaceFolder = undefined, public needRestart: boolean = false,
```
Comment (andxu) [inline src/Tomcat/TomcatServer.ts]: keep the first three arguments only private _name: string, private _tomcatPath: string, private _extensionPath: string

File: src/Tomcat/TomcatServer.ts [hunk 5]
```diff
@@ -5,20 +5,18 @@ import * as vscode from "vscode";
...
     private _debugWorkspace: vscode.WorkspaceFolder;
-    private _isDebugging: boolean = false;
-    private _state: ServerState = ServerState.IdleServer;
-
-    constructor(name: string, tomcatPath: string, extensionPath: string) {
-        this._name = name.trim();
-        this._tomcatPath = tomcatPath.trim();
-        this._extensionPath = extensionPath.trim();
+    // tslint:disable-next-line:no-parameter-properties
+    constructor(private _name: string, private _tomcatPath: string, private _extensionPath: string) {
```
Comment (jdneo) [inline src/Tomcat/TomcatServer.ts]: remove `_tomcatPath` and `_tomcatPath` if it's not used.

File: src/extension.ts [hunk 6]
```diff
@@ -142,16 +119,21 @@ async function openServerConfig(tomcatController: TomcatController, tomcatItem ?
...
-
-    if (tomcatController.getTomcatServer(serverName)) {
-        vscode.window.showInformationMessage(DialogMessage.serverExist);
-    } else {
-        await tomcatController.createTomcatServer(serverName, tomcatPath);
+    const pathPick: vscode.Uri[] = await vscode.window.showOpenDialog({
+        defaultUri: vscode.workspace.rootPath ? vscode.Uri.file(vscode.workspace.rootPath) : undefined,
+        canSelectFiles: false,
+        canSelectFolders: true,
+        openLabel: DialogMessage.selectDirectory
```
Comment (jdneo) [inline src/extension.ts]: I think we should add `canSelectMany: false` here

File: src/extension.ts [hunk 7]
```diff
@@ -162,39 +144,56 @@ async function runWarPackage(tomcatController: TomcatController, uri?: vscode.Ur
...
-    let serverPick: PickWithData<string> | undefined;
-    let serverPicks: PickWithData<string>[] = [];
-
-    if (serverSet && serverSet.length !== 0) {
-        serverPicks = serverPicks.concat(serverSet.map((server: TomcatServer) =>
-            new PickWithData(server.getName(), server.getName())));
-        if (withNew) {
-            serverPicks.push(new PickWithData(withNew, DialogMessage.createServer));
+    const pick: vscode.QuickPickItem = await vscode.window.showQuickPick(
+        [...serverSet, { label: `$(plus) ${DialogMessage.createServer}`, description: null }],
```
Comment (jdneo) [inline src/extension.ts]: better to write something to the description. not set it to 'null'

File: src/extension.ts [hunk 8]
```diff
@@ -162,39 +144,56 @@ async function runWarPackage(tomcatController: TomcatController, uri?: vscode.Ur
...
 }
 
 async function runOnTomcat(tomcatController: TomcatController, debug: boolean, uri?: vscode.Uri): Promise<void> {
-    const ui: VSCodeUI = new VSCodeUI();
-    const packagePath: string = uri ? uri.fsPath : await ui.showFileFolderDialog(true, false, DialogMessage.selectWarPackage);
+    if (!uri) {
+        const dialog: vscode.Uri[] = await vscode.window.showOpenDialog({
+            defaultUri: vscode.workspace.rootPath ? vscode.Uri.file(vscode.workspace.rootPath) : undefined,
+            canSelectFiles: true,
+            canSelectFolders: false,
```
Comment (jdneo) [inline src/extension.ts]: same here: add `canSelectMany: false`

File: src/extension.ts (hunk 7 above)
Comment (SummerSun) [reply src/extension.ts]: description will show up in the dialog, not what we are expecting. Or we could set it empty.

File: src/extension.ts (hunk 8 above)
Comment (SummerSun) [reply src/extension.ts]: canSelectMany default false. Scenario tested.

File: src/Tomcat/TomcatServer.ts (hunk 5 above)
Comment (SummerSun) [reply src/Tomcat/TomcatServer.ts]: #1: those two were used currently.  will betternaming it or centralized the using to controller in following PR.

#2: good suggestion.

**Pull request merged**